deepsubtree.go

```diff
@@ -423,6 +331,44 @@ func (dst *DeepSubTree) recursiveRemove(node *Node, key []byte) (newHash []byte,
 	return nil, nil, nil, fmt.Errorf("node with key: %s not found", key)
 }
 
+func (tree *MutableTree) getExistenceProofsNeededForSet(key []byte, value []byte) ([]*ics23.ExistenceProof, error) {
+	tree.Set(key, value)
+
+	return tree.getExistenceProofsNeeded()
+}
+
+func (tree *MutableTree) getExistenceProofsNeededForRemove(key []byte) ([]*ics23.ExistenceProof, error) {
+	ics23proof, err := tree.GetMembershipProof(key)
+	if err != nil {
+		return nil, err
+	}
+
+	tree.Remove(key)
+
+	existenceProofs, err := tree.getExistenceProofsNeeded()
+	if err != nil {
+		return nil, err
+	}
+	existenceProofs = append(existenceProofs, ics23proof.GetExist())
+	return existenceProofs, nil
+}
+
+func (tree *MutableTree) getExistenceProofsNeeded() ([]*ics23.ExistenceProof, error) {
+	keysAccessed := tree.ndb.keysAccessed.Values()
+
+	tree.Rollback()
```

Comment:
This is quite "unexpected" side-effect.

Reply:
factored it out so it exists in remove and set individually